### PR TITLE
Adding noFilesWithArrayNotation option

### LINKED ISSUE #93: Multiple Files in ASP.NET Core
This is either an bug of a different implementation, but converting an array of `File` objects doesn't work in ASP.NET Core. 

For example:
```js
var model = {
	battlePlans: [new File(), new File()]
};
const formData = objectToFormData(model);
console.log(formData);

// Currently returns:
// battlePlans[]: (binary)
// battlePlans[]: (binary)
```

But apparently ASP.NET Core expects the following:
```js
var model = {
	battlePlans: [new File(), new File()]
};
const formData = objectToFormData(model);
console.log(formData);

// Expected to return:
// battlePlans: (binary)
// battlePlans: (binary)
```
_Note the missing `[]`._

The only reference from **Microsoft** saying that it should be done like this, is in [the following section](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-3.1#file-upload-scenarios):
image
The [note reference](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-3.1#match-name-attribute-value-to-parameter-name-of-post-method) to the following code example:
```js
var formData = new FormData();
for (var file in files) {
  formData.append("battlePlans", file, file.name);
}
```

**MDN** on the other side shows an [example](https://developer.mozilla.org/en-US/docs/Web/API/FormData/append#Example) where they use the brackets, saying that this is the way to be compatible with PHP:
```js
formData.append('userpic[]', myFileInput.files[0], 'chris1.jpg');
formData.append('userpic[]', myFileInput.files[1], 'chris2.jpg');
```

**WHATWG** is _not_ specific about [this](https://xhr.spec.whatwg.org/#concept-formdata-entry-list).

This appears to be a problem only for an array of `File` objects. Other types work just fine with current implementation.

### PR DESCRIPTION
**Pull request by jerone**
Closes #93.

### CODE REVIEW

**Comment by jerone** [comment]
@therealparmesh Anything else I need to do?

**Comment by trueboris** [comment]
@therealparmesh is there anything needed for this PR to be merged?

**Comment by isaacfilarski** [comment]
Please merge 
297nvg

File: README.md
Comment (therealparmesh) [inline README.md:48]: There will be a configuration option for dot notation instead of bracket notation. Can we name this flag something else to make it sound more compatible?

Comment (jerone) [reply README.md:48]: @therealparmesh I'm not sure which new configuration you are talking about...

This option is only about a list of `File` types. How about `noFileCollectionBrackets`?

**Comment by jerone** [comment]
@[therealparmesh](https://github.com/therealparmesh/object-to-formdata/commits?author=therealparmesh) Before I fix the merge conflicts, are you still interested in this?

**Comment by MSerj** [comment]
useful option, please merge 👍

**Comment by Itsamirmasoud** [comment]
@jerone hey man....could you please take a look at this PR ? this is a necessary bug-fix for me. thanks for the great job

**Comment by jerone** [comment]
@Itsamirmasoud What can I do for you? This PR is all ready for merging.

**Comment by Itsamirmasoud** [comment]
> @Itsamirmasoud What can I do for you? This PR is all ready for merging.
well that's the thing....merging. I was wondering how long does it take for this PR to be merged into master and the new version released on npm?

**Comment by jerone** [comment]
@Itsamirmasoud As you can see, this is my PR (not my repo); I'm waiting for approval and merging. You have to ask @therealparmesh, as I'm waiting for that too.

**Comment by Itsamirmasoud** [comment]
> @Itsamirmasoud As you can see, this is my PR (not my repo); I'm waiting for approval and merging. You have to ask @therealparmesh, as I'm waiting for that too.

🤦‍♂️sorry ... my bad.

**Comment by Itsamirmasoud** [comment]
@jerone 
so I tested this on my project(.net core 5 & vue 2 + axios) since i could not wait for the merge  and it turned out that this does not work.

Did you have any other configuration for model binder or anything else?

I have a complex model which look like this : 

```
    public class Test
    {
        public class TestFile
        {
            public byte FileTypeId { get; set; }
            public IFormFile File { get; set; }
        }

        public int SomeOtherId { get; set; }
        public List<TestFile> Files { get; set; }
    }

```
and my JS object look like this : 

```
const result = [{  FileTypeId: this.fileTypeId, file },{...},{...}];

const data  = { files: result, someOtherId: this.someOtherId};

const formData = serialize(data );
```

and although the noFileListBrackets was set to true it still produced

```files[0][file] = [file object] ```
which the .net core model binder could not understand.

so after debugging it ...i changed this piece of code :
```
      const key = pre ? pre + '[' + prop + ']' : prop;
```

to this : 

```
            let key;

            if (cfg.noFileListBrackets && isFile(value)) {
                key = pre ? pre + '.' + prop : prop;
            } else {
                key = pre ? pre + '[' + prop + ']' : prop;
            }
```

and now it works.

**Comment by therealparmesh** [comment]
We should come up with a plan to merge this in.

First, we should try to think of all test cases and represent them here.

Then we should think about what to call the flag. Remember that bracket notation (`foo[bar]`) isn't the only valid notation. Dot notation (`foo.bar`) is likely going to need to be added in the near future. I think we should brainstorm another flag name, since `noFileListBrackets` wouldn't make sense in that (likely) scenario.

**Comment by jerone** [comment]
@therealparmesh What about naming it `noPropertiesForFileList` or something related. No matter if it stays at bracket notation or will be dot notation, ASP.NET requires the plain property (with or without indices).

@Itsamirmasoud My environment is a little different, I'm running Angular on the front-end and ASP.NET Core 3.1 on the back-end. Aldo ASP.NET Core 5 still [suggests the same configuration](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-5.0#match-name-attribute-value-to-parameter-name-of-post-method).

**Comment by Itsamirmasoud** [comment]
> @Itsamirmasoud My environment is a little different, I'm running Angular on the front-end and ASP.NET Core 3.1 on the back-end. Aldo ASP.NET Core 5 still [suggests the same configuration](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-5.0#match-name-attribute-value-to-parameter-name-of-post-method).
> Are you sure your datatype for `files` is `FileList`?

I am not quite sure I understand your point by `FileList`.

`files` is an array of JS objects in the frontend : 

```
const files= [{  FileTypeId: this.fileTypeId, file },{...},{...}];
```

 and a matching C# class is defined for it in backend :

```
public class TestFile
{
      public byte FileTypeId { get; set; }
      public IFormFile File { get; set; }
}

public List<TestFile> Files { get; set; }
```

**Comment by jerone** [comment]
@Itsamirmasoud Sorry, I'm into TypeScript, which has an interface called `FileList` ([link](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.filelist.html)), which is an array of files (`File[]`) ([link](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.file.html)).

It does like your array is incorrect. It should be only an array of files. Not an array of objects, which contains a file.

Like this:

```
const files = [ File, File, File, ... ];
```

And the file should comply to the following properties to be recognized as a file: https://github.com/therealparmesh/object-to-formdata/blob/9d5fc4a435fb13b4622936667881f2b534d2231c/src/index.js#L13-L23

**Comment by Itsamirmasoud** [comment]
> @Itsamirmasoud Sorry, I'm into TypeScript, which has an interface called `FileList` ([link](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.filelist.html)), which is an array of files (`File[]`) ([link](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.file.html)).
> 
> It does like your array is incorrect. It should be only an array of files. Not an array of objects, which contains a file.
> 
> Like this:
> 
> ```js
> const files = [ File, File, File, ... ];
> ```
> 
> And the file should comply to the following properties to be recognized as a file:
> 
> https://github.com/therealparmesh/object-to-formdata/blob/9d5fc4a435fb13b4622936667881f2b534d2231c/src/index.js#L13-L23
> 
> _Edit:_
> 
> In C# you can use the interface `IFormFileCollection` for your files. You'll lose the `FileTypeId`, but that can be an list too.
> 
> ```cs
> public class MyRequest {
>   public IFormFileCollection Files { get; set; }
>   public IEnumerable<byte> FileTypeIds { get; set; }
>   public string OtherProperties { get; set; }
> }
> ```
Well the files i am putting in my object are of `File` type you mentioned because i am using typescript too  :) , though i did not know about `FileList` interface.

About `IFormFileCollection `.... i have tried this before and yes it works if you are posting only an array of files like you mentioned.
but my case is different, the `FileTypeId` and the `File` are a pair  and it is not possible to omit any of them.

Apart from all this, I managed to get it working with the solution I mentioned in the above comment and i think it is what it is. at the end of day that's how .Net 5 model binder works and other libraries has to adapt to it.

**Comment by jerone** [comment]
Ah ok, then your issue is somewhat related, but different. This PR is about `File` lists (e.g. `FileList`). Your issue is about a `File` within an `object` within an `array`.

You are looking to change this line to your suggestion ?! CT-Energy/object-to-formdata@[hotfix/%2393-support-multiple-files/src/index.js#L81](https://github.com/CT-Energy/object-to-formdata/blob/hotfix/%2393-support-multiple-files/src/index.js#L81)

If you can send me an [PR with included test](https://github.com/CT-Energy/object-to-formdata/tree/hotfix/%2393-support-multiple-files), I can include it in this PR.

**Comment by Itsamirmasoud** [comment]
> Ah ok, then your issue is somewhat related, but different. This PR is about `File` lists (e.g. `FileList`). Your issue is about a `File` within an `object` within an `array`.
> 
> You are looking to change this line to your suggestion ?! CT-Energy/object-to-formdata@[hotfix/%2393-support-multiple-files/src/index.js#L81](https://github.com/CT-Energy/object-to-formdata/blob/hotfix/%2393-support-multiple-files/src/index.js#L81)
> 
> If you can send me an [PR with included test](https://github.com/CT-Energy/object-to-formdata/tree/hotfix/%2393-support-multiple-files), I can include it in this PR.

ok, thanks i will try to send you a PR.

**Comment by code-sumedh** [comment]
Any plans to merge this PR?

**Comment by tsafadi** [comment]
This should be merged asap. The util is not useful without this feature and is fairly simple to integrate.
Asp.Net is one of the most used framworks for backend dev.

**Comment by jerone** [comment]
I've been running this fix in my [fork](https://github.com/CT-Energy/object-to-formdata/tree/hotfix/%2393-support-multiple-files) for over an year, without any issues.

:shipit:

**Comment by tsafadi** [comment]
Why is it not being merged? @jerone

**Comment by jerone** [comment]
I don't know 🤷‍♂️

**Comment by tsafadi** [comment]
What is the reason this fork is not being merged? Anything else needed? Please give us an update @therealparmesh

**Comment by therealparmesh** [comment]
It hasn't been merged due to my personal life taking priority recently. I'll get back to this soon. Thank you for the feedback.

**Comment by jerone** [comment]
🏓

**Comment by fvandillen** [comment]
This is taking too long for a simple bugfix. Why can't this be merged?

**Comment by therealparmesh** [comment]
> This is taking too long for a simple bugfix. Why can't this be merged?

Don't be a jackass.

**Comment by therealparmesh** [comment]
@jerone Thank you for your amazing work and patience here. I hope you don't mind, but I added some changes on top of yours in order to keep this PR up-to-date, to change the configuration name to make it future-proof, and to add more test cases.

I propose changing the name to `noFilesWithArrayNotation` because there is a likely future need to be able to change from using brackets to using periods (or anything else) to notate an object or an array in the key. I feel strongly that the name should be agnostic to these kind of details.

I also added an edge case to the new test, which I think is important to think about as we add more configuration options.

I put these changes are here: https://github.com/therealparmesh/object-to-formdata/compare/%2393

Please let me know your thoughts, and please bring those changes in if you agree with them.

**Comment by jerone** [comment]
We have been using these changes in production since the opening of this PR, so I can safely say the change works. 😀

I have no opinion about the naming; yours is perfectly fine too.

I'll pull in the changes from your branch.

**Comment by jerone** [comment]
Changes merged, should be the same as your branch.

https://github.com/therealparmesh/object-to-formdata/compare/ce2cde5dab635e463141e17d85a0085ac7a9959a...%2393

**Review by therealparmesh — APPROVED** [review]

**Comment by therealparmesh** [comment]
Merged and published.

**Pull request merged**
